File: AudioRecorder2/experiments/cider_utils.js

function seq(start,end) { return [...Array(end-start+1).keys()].map((x) => start+x) }

const stddev = function (data) { return new StatsHandler(data).stddev() };

const quantile = function (arr, q) { return new StatsHandler(arr).quantile(q) };

const stats = function(arr) { return new StatsHandler(arr).stats() }

let DisableAllPrinting = false

function p(msg) { if (!DisableAllPrinting) { console.log(msg) } }

class StatsHandler {

  constructor(name, maxSize = 500, skips = 10000) {
    this.name = name
    this.maxSize = maxSize
    this.skips = skips
    this.cntr = 0
    this.data = []
  }

  mean = (arr=this.data) => arr.reduce((acc, val) => acc + val, 0) / arr.length;

  stddev = function (data=this.data) {
    if (!data || data.length <= 1) {
      return 0
    } else {
      let m = this.mean(data);
      return Math.sqrt(data.reduce(function (sq, n) {
        return sq + Math.pow(n - m, 2);
      }, 0) / (data.length - 1));
    }
  };

  quantile(q, arr=this.data) {
    let sorted = [...arr]
    sorted.sort((a, b) => a - b)
    return sorted[Math.floor(q * (arr.length - 1))]
  }

  /** Use with care: keeps the largest values */
  quartileKeepLargest(arr, q) {
    let sorted = arr.sort( (a,b)=> a-b)
    return sorted[ Math.floor(q * (arr.length-1))]
  }

  stats(arr = this.data) {
    return {
      counter: this.cntr,
      bufSize: this.data.length,
      mean: this.mean(arr),
      stdev: this.stddev(arr),
      min: Math.min(...arr),
      max: Math.max(...arr),
      '1%': this.quantile( .01, arr),
      '10%': this.quantile( .1, arr),
      median: this.quantile( .5, arr),
      '90%': this.quantile( .9, arr),
      '99%': this.quantile( .99, arr)
    }
  }

  statsToString(arr=this.data) {
    return JSON.stringify(arr, null, 2)
  }

  toLogString() {
    return `[${Date.now()} ${this.name}-Stats ${this.cntr}] ${this.statsToString(this.stats())}`
  }

  print() {
    p(this.toLogString())
  }

  add(stat) {
    if (DefaultConf.enableStats) {
      if (this.data.length >= this.maxSize) {
        this.data.shift()
      }
      this.data.push(stat)
      this.cntr++
      if (this.cntr % this.skips == 1) {
        this.print()
      }
    }
  }
}

class Md5Manager {

  constructor(name, maxBufsCnt=256) {
    this.name = name
    this.MaxBufsCnt = maxBufsCnt
    this.bufs = []
    this.bufs2 = []
    this.stats = new StatsHandler(`${name}.`,500,10000)
  }

  printStats() {
    p(this.stats.toLogString())
  }

  printMd5(msg,mdd) {
    p(`Stats-${this.name} ${msg} ${mdd}`)
  }

  add(buf) {
    if (DefaultConf.enableMd5) {
      this.stats.add()
      let md55 = md5(buf)
      this.bufs.push(md55)
      if (this.bufs.length == this.MaxBufsCnt) {
        let md5b = md5(this.bufs.join(''))
        this.printMd5(`Level1'-${name}:`, md5b)
        this.bufs = []
        this.bufs2.push(md5b)
        if (this.bufs2.length == this.MaxBufsCnt) {
          let md5b = md5(this.bufs2.join(''))
          this.printMd5(`Level2'-${name}:`, md5b)
          this.bufs2 = []
        }
      }
    }
  }
}

function dcopy(x) {
	return JSON.parse(JSON.stringify(x))
}

function jprint(j) { p(JSON.stringify(j)) }

const l2Norm = function (data) {
    return Math.sqrt(data.reduce(function (sq, n) {
        return sq + Math.pow(n, 2);
    }, 0) / (data.length - 1));
};

function partition(arr,len) {
  return [...new Array(Math.ceil(arr.length / len))].map(_ => arr.splice(0, len))
}

function floatArrToArr(floatArr) {
  return [...Array(floatArr.length)].map(
    (x,i) => floatArr[i]
  )
}

function arrToFloatArr(arr) {
  let floatArr = new Float32Array(arr.length);
  [...arr].map( (x,i) => floatArr[i] = x)
  return floatArr
}


/** Multithreading not currently supported */
class PerfManager {
  constructor(name, maxSize=1000,printSkips=1000) {
    this.starts = {} // {str:str}
    this.stats = {} // {str:StatsHandler}
    this.maxSize = maxSize
    this.printSkips = printSkips
  }

  start(name) {
    this.starts[name] = Date.now()
  }
  stop(name) {
    if (DefaultConf.enablePerf) {
      if (this.stats[name] === undefined) {
        this.stats[name] = new StatsHandler(`perf.${name}`, this.maxSize, this.printSkips)
      }
      this.stats[name].add(Date.now() - this.starts[name])
    }
  }

  toString() {
    return `Perf-${this.name} ${this.stats.toLogString()}`
  }

}

File: AudioRecorder2/experiments/CiderUtils.swift
//
// Created by steve on 6/14/20.
// Copyright (c) 2020 Yaroslav Zhurakovskiy. All rights reserved.
//

import Foundation

let DisableAllPrinting = false

func p(_ msg: String) {
    if (!DisableAllPrinting) {
        print(msg)
    }
}

struct Stats: CustomDebugStringConvertible, CustomStringConvertible {
    let counter: Int
    let bufSize: Int
    let mean: Float
    let stdev: Float
    let min: Float
    let max: Float
    let p1: Float
    let p10: Float
    let median: Float
    let p90: Float
    let p99: Float

    init(counter: Int, bufSize: Int, mean: Float, stdev: Float, min: Float, max: Float, p1: Float, p10: Float, median: Float, p90: Float, p99: Float) {
        self.counter = counter
        self.bufSize = bufSize
        self.mean = mean
        self.stdev = stdev
        self.min = min
        self.max = max
        self.p1 = p1
        self.p10 = p10
        self.median = median
        self.p90 = p90
        self.p99 = p99
    }

    var description: String {
        "Stats(counter: \(counter), bufSize: \(bufSize), mean: \(mean), stdev: \(stdev), min: \(min), max: \(max), p1: \(p1), p10: \(p10), median: \(median), p90: \(p90), p99: \(p99))"
    }
    var debugDescription: String {
        description
    }
}

class StatsHandler {

    public static var EnableStats = true
    typealias Num = Float32
    typealias Data = [Num]
    let name: String
    let maxSize: Int
    var skips: Int
    var cntr = 0
    var data: Data = []
    // TODO: change to Deque for better performance
//    let data: Deque<Num>

    init(_ name: String, _ maxSize: Int, _ skips: Int) {
        self.name = name
        self.maxSize = maxSize
        self.skips = skips
//        self.data =  Deque<Num>(minimumCapacity: maxSize)
    }

    func mean() -> Num {
        let arr = data
        return arr.reduce(0, { acc, val in acc + val }) / Float(arr.count)
    }

    func stddev() -> Num {
        if (data.count <= 1) {
            return 0
        } else {
            let m = self.mean();
            return sqrt(data.reduce(0, { sq, n in
                return sq + pow(n - m, 2)
            }) / Float(data.count - 1));
        }
    }

    func quantile(_ q: Float) -> Float {
        let arr = self.data
        let sorted = arr.sorted()
//    sorted.sort((a, b) => a - b)
        return sorted[Int(floor(q * (Float(arr.count) - 1)))]
    }


    func stats() -> Stats {
        let arr = self.data
        return Stats(
                counter: self.cntr,
                bufSize: self.data.count,
                mean: self.mean(),
                stdev: self.stddev(),
                min: arr.min()!,
                max: arr.max()!,
                p1: self.quantile(0.01),
                p10: self.quantile(0.1),
                median: self.quantile(0.5),
                p90: self.quantile(0.9),
                p99: self.quantile(0.99)
        )
    }

    public func statsToString() -> String {
        return stats().debugDescription
    }

    public func toLogString() -> String {
        return "[\(now())] \(self.name)-Stats \(self.cntr)] \(self.statsToString())"
    }

    func print() {
        p(toLogString())
    }

    public func push(_ stat: Num) {
        if (Self.EnableStats) {
            if (self.data.count >= self.maxSize) {
                self.data.remove(at: self.data.count)
            }
            data.insert(stat, at: 0)
            self.cntr += 1
            if (self.cntr % self.skips == 1) {
                self.print()
            }
        }
    }
}

func now() -> Int {
    return Int(Date().timeIntervalSince1970 * 1000)
}

func formattedDate() -> String {
    let fmt = DateFormatter()
    fmt.dateFormat = "MM-dd HH24:mm:ss.SSSS"
    fmt.timeZone = TimeZone.current
//    return "\(Date().getFormattedDateString(format: "yyyy-MM-dd HH:mm:ss"))"
    return fmt.string(from: Date())
}


class PerfManager {
    public static let EnablePerf = true
    var starts: [String: Int] = [:]
    var stats: [String: StatsHandler] = [:]
    let maxSize: Int
    let printSkips: Int

    init(maxSize: Int = 1000, printSkips: Int = 1000) {
        self.maxSize = maxSize
        self.printSkips = printSkips
    }

    func start(name: String) {
        self.starts[name] = now()
    }

    func stop(name: String) {
        if (Self.EnablePerf) {
            if (self.stats[name] == nil) {
                self.stats[name] = StatsHandler("perf.\(name)", self.maxSize, self.printSkips)
            }
            self.stats[name]!.push(Float(now() - self.starts[name]!))
        }
    }

    func toString() -> String {
        let ms = stats.map{ k,v in "\(k)=> \(v.statsToString())"}.joined(separator: "\n")
        return "Perf - \(ms)"
    }

}